enhancements/cluster-logging/output-parameter-templates.md

```diff
@@ -0,0 +1,153 @@
+---
+title: output-parameter-templates
+
+authors:
+  - "@alanconway"
+
+reviewers:
+  - "@ewolinetz"
+  - "@lukas-vlcek"
+  - "@jcantril"
+  - "@sichvoge"
+  - "@vimalk78"
+
+approvers:
+  - "@jcantril"
+
+creation-date: 2021-04-21
+
+last-updated:
+
+status: provisional
+
+see-also:
+  - "/enhancements/cluster-logging/forwarding-json-structured-logs.md"
+---
+
+# Output Parameter Templates
+
+## Release Sign-off Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The ClusterLogForwarder.Outputs API can already set some output parameters from log record fields.
+
+For example, to set the Elasticsearch index for each log record to the value of kubernetes label "foo" on that record:
+
+``` yaml
+  elasticsearch:
+    indexKey: kubernetes.label.foo
+```
+
+This proposal allows multiple record fields to be _combined_ into an output value.
+
+For example, to set an index of the form "label_value+container_name":
+
+``` yaml
+  elasticsearch:
+    indexPattern: '{.kubernetes.label.foo}+{.kubernetes.container_name}'
```

Comment:
This is a dangerous offering for Elasticsearch as it creates a single index per container.  Imagine on a large OCP cluster with short lived build pods, crons, etc,  you will have significant shard explosion if you allow each container to have it's own index.  Additionally, modifying the shard replication leads to the need to potentially have significantly large ES clusters.  Customer's may want to do this not realizing the implications but I would N'ack this for our managed offering

Reply:
+1. It is a bad example. I'll remove and make sure we focus on realistic/useful examples.

Reply:
It has the potential for abuse, but container names are not randomized strings like pod names, they tend to name the service in the container, so there will likely be only a small set of container names across the cluster, and a very small set in the application identified by a label. 
I see no other way to solve the problem of logging multi-container pods to ES, since sidecar containres are very likely to have different logging formats from the app container - they're usually created by completely unrelated teams. E.g. the Istio project likely has nothing to do with the local application project.  
Thoughts?

Reply:
@igor-karpukhin @periklis @lukas-vlcek  Do we have enough metrics and alerts in place to provide enough visibility into the issues I foresee this change introducing?

Reply:
@jcantrill good point. Note this is also relevant to the changes we have already made - if the user picks a label with too many values in the cluster they can also overload ES. This whole user-chosen index thing is inherently prone to excessive ES indices if the user isn't careful.  That's been a concern since the start, but we don't have any better ideas (so far as I know) to prevent a single index from exploding due to incompatible JSON messages.

Reply:
@jcantrill can you lgtm this as an initial proposal? We can add details about metrics if need be later - that applies in any case to what we have already implemented and shipped.